### DISCUSSION TITLE
Can a uniform buffer be updated after creation using map_async?

### DISCUSSION THREAD

**Discussion by shi-yan** [gfx-rs/wgpu#3357] [Q&A] [answered]
I have a uniform buffer containing a projection matrix.

As I move my camera, I need to update this matrix. 

I tried using map_async on that uniform buffer, but I received errors. I can't seem to be able to combine UNIFORM with MAP_WRITE and COPY_DST

it seems that a uniform buffer can't be updated after creation. Am I right?

**Comment by cwfitzgerald** [comment, marked as answer]
While you _kinda_ can do this, the preferred way to update data is by using queue.write_buffer

**Reply by shi-yan** [reply]
How? UNIFORM flag can't be used with MAP_WRITE

**Reply by cwfitzgerald** [reply]
There's a native extension that allows this. That being said it's a performance footgun and you shouldn't use it. If you wanted to deal with uploads yourself, you can write your data into a staging buffer with MAP_WRITE | COPY_SRC, then issue a copy to the real buffer. This is how write_buffer works internally.

**Reply by shi-yan** [reply]
Thank you so much! One more question, if you don't mind, what's the reason behind UNIFORM's incompatibility with MAP_WRITE? I read somewhere that the uniform buffer is on the device's local memory. Is that true?

**Reply by cwfitzgerald** [reply]
Yeah, the good tl;dr is that everything with MAP in it is in CPU memory, everything without is in GPU memory.
